File: group-anagrams/njngwn.java
// Time Complexity: O(n), n: strs.length
// Space Complexity: O(m+n), m: strMap.size, n: anagramList.size
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, ArrayList<String>> strMap = new HashMap<String, ArrayList<String>>();

        for (String str : strs) {
            char[] charArr = str.toCharArray();
            Arrays.sort(charArr);
            String key = String.valueOf(charArr);

            if (strMap.containsKey(key)) {
                strMap.get(key).add(str);
            } else {
                strMap.put(key, new ArrayList<String>(Arrays.asList(str)));
            }
        }

        return new ArrayList<>(strMap.values());
    }
}

File: group-anagrams/njngwn.py
class Solution:
    # Time Complexity: O(n), n: len(strs)
    # Space Complexity: O(n), n: len(strs)
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagram_groups = dict()  # key: string in an alphabetical order, values: strings

        for s in strs:
            word = ''.join(sorted(s))
            if word in anagram_groups:
                anagram_groups[word].extend([s])
            else:
                anagram_groups[word] = [s]

        return list(anagram_groups.values())
